src/rust/engine/BUILD

```diff
@@ -0,0 +1,36 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+files(
+    name="rust_sources",
+    sources=[
+        "Cargo.lock",
+        "build.rs",
+        "**/Cargo.toml",
+        "**/*.rs",
+        "**/*.proto",
+        ".cargo/config",
+        "rust-toolchain",
+    ],
+)
+
+shell_sources(name="shell")
+
+shell_command(
+    name="engine-and-client",
+    command=f"./cargo_build_shim.sh --features=extension-module -p engine -p client",
```

Comment:
OK last comment :smile: 

I tested this and it does seem you get a full recompile each time, which of course, isn't ideal. It'd be great if we could stash the cache someone outside the sandbox, but of course that comes with the usual warnings :smile:

Reply:
I'm thinking https://lib.rs/crates/sccache might be our way out of full recompilation.

Building the `engine` itself (likely due to linking?) is still slow.

Reply:
> I tested this and it does seem you get a full recompile each time, which of course, isn't ideal. It'd be great if we could stash the cache someone outside the sandbox, but of course that comes with the usual warnings 😄

Yeah, having `shell_command`/`adhoc_tool` able to point to point to a `named_cache/...` could be handy for this sort of thing... although the details of getting that to work nicely are obviously up in the air.

> I'm thinking https://lib.rs/crates/sccache might be our way out of full recompilation.

Yeah, I would think sccache is a good first step to improve this. 👍 

... Although, the current `tools` config here means we _require_ a user to have `sccache` installed, and, the invocation of cargo here doesn't actually use it yet. In addition, I notice in https://github.com/mozilla/sccache/blob/main/docs/Local.md that they say "The local storage only supports a single sccache server at a time. Multiple concurrent servers will race and cause spurious build failures". I wonder whether running sccache in the pants sandbox will end up with concurrent servers.

Thus, maybe we start _without_ sccache for now and work through adding it in incrementally after this PR?

Reply:
> Thus, maybe we start without sccache for now and work through adding it in incrementally after this PR?

Yeah, IDK what you're seeing but this PR doesn't introduce `sccache` yet :eyes:

Reply:
The `tools` line includes `sccache` as the last item, at the moment. https://github.com/pantsbuild/pants/blob/9cd6fd6e09273879835ff6412dbdee73891c472a/src/rust/engine/BUILD#L29